workspaces/arborist/lib/query-selector-all.js

```diff
@@ -422,6 +431,135 @@ class Results {
   dedupedPseudo () {
     return this.initialItems.filter(node => node.target.edgesIn.size > 1)
   }
+
+  async outdatedPseudo () {
+    const { outdatedKind = 'any' } = this.currentAstNode
+
+    // filter the initialItems
+    // NOTE: this uses a Promise.all around a map without in-line concurrency handling
+    // since the only async action taken is retrieving the packument, which is limited
+    // based on the max-sockets config in make-fetch-happen
+    const initialResults = await Promise.all(this.initialItems.map(async (node) => {
+      // the root can't be outdated, skip it
+      if (node.isProjectRoot) {
+        return false
+      }
+
+      // we cache the promise representing the full versions list, this helps reduce the
+      // number of requests we send by keeping population of the cache in a single tick
+      // making it less likely that multiple requests for the same package will be inflight
+      if (!this.#outdatedCache.has(node.name)) {
+        this.#outdatedCache.set(node.name, getPackageVersions(node.name, this.flatOptions))
+      }
+      const availableVersions = await this.#outdatedCache.get(node.name)
```

Comment:
nit: awaiting is "expensive" and if we `set` the awaited value we are only doing it once.

Reply:
in my admittedly limited testing, the overhead of awaiting this value repeatedly was vastly smaller than the repeated work done by awaiting a result before storing it in the cache (~4.5 seconds awaiting before storing, ~3.1 seconds storing before awaiting on my machine in this repo)

that being the case, i think i'd like to leave this as-is for now and we can always tweak it later

Reply:
boom. resolved.